feat: Implement client-side validation and handle async searchParams

dfdumaresq opened the pull request:
This commit introduces client-side validation for the signup form, enhancing the user experience by providing real-time feedback for email and password inputs. It also addresses the issue of handling the `searchParams` prop, which is a `Promise<Message>`, by resolving it using `useEffect` and `useState`.

Key changes:

- Implemented email and password validation functions with real-time error messages.
- Added state management for email, password, and form validity.
- Disabled the submit button until the form is valid.
- Resolved the `Promise<Message>` using `useEffect` and stored the result in state.
- Displayed a loading indicator while waiting for `searchParams` to resolve.
- Rendered the `FormMessage` component conditionally after `searchParams` has been resolved.

This ensures that the component handles asynchronous data correctly, provides a smoother user experience, and improves type safety.

dfdumaresq committed:
feat: Implement client-side validation and handle async searchParams

This commit introduces client-side validation for the signup form, enhancing the user experience by providing real-time feedback for email and password inputs. It also addresses the issue of handling the `searchParams` prop, which is a `Promise<Message>`, by resolving it using `useEffect` and `useState`.

Key changes:

- Implemented email and password validation functions with real-time error messages.
- Added state management for email, password, and form validity.
- Disabled the submit button until the form is valid.
- Resolved the `Promise<Message>` using `useEffect` and stored the result in state.
- Displayed a loading indicator while waiting for `searchParams` to resolve.
- Rendered the `FormMessage` component conditionally after `searchParams` has been resolved.

This ensures that the component handles asynchronous data correctly, provides a smoother user experience, and improves type safety.

pszjmb1 reviewed (changes requested):
Hi,

Thank for the changes. I noticed a couple of niggles that ought to be fixed before merging:

1.  Line 37 has `const minLength = 8;` but this conflicts with line 122 `minLength={6}`. So when setting the password the message isn't correct. And then if the password is less than 12 characters this message appears after clicking signup:

![Untitled](https://github.com/user-attachments/assets/6c82fedf-8a36-405d-a203-8186c54389f9)

dfdumaresq committed:
fix: improve form message visibility and styling

- Add contrasting background colors for different message types
- Implement distinct border styling for better visual hierarchy
- Add proper padding and rounded corners for message containers
- Fix message visibility issues in FormMessage component

dfdumaresq committed:
- Create pre-build script to add required variable to .env.local: NEXT_PUBLIC_MIN_PASSWORD_LENGTH
  - Update package.json with prebuild command and @iarna/toml to assist in parsing value from config.toml.

dfdumaresq commented:
> Hi,
> 
> Thank for the changes. I noticed a couple of niggles that ought to be fixed before merging:
> 
> 1. Line 37 has `const minLength = 8;` but this conflicts with line 122 `minLength={6}`. So when setting the password the message isn't correct. And then if the password is less than 12 characters this message appears after clicking signup:
> 

Hi Jessie, 
I'm pulling the minLength now from .env.local and using it as a const for validation in `Signup`. 
`const minLength = Number(process.env.NEXT_PUBLIC_MIN_PASSWORD_LENGTH) || 12;`
```
...
          <Input
            type="password"
            name="password"
            placeholder="Your password"
            value={password}
            onChange={handlePasswordChange}
            minLength={ minLength }
            required
          />
```
Also fixed a length comparison that made the validation numerically off by 1, which should fix the issue you're seeing about the length being less than the number required.

`    if (value.length < minLength-1) {`

Finally, I had to add some styling to `form-message.tsx` because the validation messages were not visible on my browser (Brave). Hope this hasn't broken anything...
